x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/TRange.java

```diff
@@ -0,0 +1,305 @@
+            FIRST,
+            "string",
+            "long",
+            "date",
+            "date_nanos"
+        ).and(isType(second, dt -> dt == first.dataType(), operationName, SECOND, first.dataType().esType()));
+
+        if (resolution.unresolved()) {
+            return resolution;
+        }
+
+        return TypeResolution.TYPE_RESOLVED;
+    }
+
+    @Override
+    public Expression replaceChildren(List<Expression> newChildren) {
+        if (newChildren.size() == 3) {
+            return new TRange(source(), newChildren.get(0), newChildren.get(1), newChildren.get(2), configuration());
+        }
+        return new TRange(source(), newChildren.get(0), newChildren.get(1), null, configuration());
+    }
+
+    @Override
+    public Expression surrogate() {
+        long[] range = getRange(FoldContext.small());
+
+        Expression startLiteral = new Literal(source(), range[0], timestamp.dataType());
+        Expression endLiteral = new Literal(source(), range[1], timestamp.dataType());
+
+        return new And(source(), new GreaterThan(source(), timestamp, startLiteral), new LessThanOrEqual(source(), timestamp, endLiteral));
+    }
+
+    @Override
+    protected NodeInfo<? extends Expression> info() {
+        return NodeInfo.create(this, TRange::new, timestamp, first, second, configuration());
+    }
+
+    @Override
+    public Nullability nullable() {
+        return timestamp.nullable();
+    }
+
+    private long[] getRange(FoldContext foldContext) {
+        Instant rangeStart;
+        Instant rangeEnd;
+
+        try {
+            Object foldFirst = first.fold(foldContext);
+            if (second == null) {
+                rangeEnd = configuration().now().toInstant();
+                rangeStart = timeWithOffset(foldFirst, rangeEnd);
+            } else {
+                Object foldSecond = second.fold(foldContext);
+                rangeStart = parseToInstant(foldFirst, START_TIME_OR_OFFSET_PARAMETER);
+                rangeEnd = parseToInstant(foldSecond, END_TIME_OR_OFFSET_PARAMETER);
+            }
+        } catch (InvalidArgumentException e) {
+            throw new InvalidArgumentException(e, "invalid time range for [{}]: {}", sourceText(), e.getMessage());
+        }
+
+        if (rangeStart.isAfter(rangeEnd)) {
+            throw new InvalidArgumentException("TRANGE rangeStart time [{}] must be before rangeEnd time [{}]", rangeStart, rangeEnd);
+        }
+
+        if (timestamp.dataType() == DataType.DATE_NANOS && first.dataType() == DataType.DATE_NANOS) {
+            return new long[] { DateUtils.toLong(rangeStart), DateUtils.toLong(rangeEnd) };
+        }
+
+        boolean convertToNanos = timestamp.dataType() == DataType.DATE_NANOS;
+        return new long[] {
+            convertToNanos == false ? rangeStart.toEpochMilli() : DateUtils.toNanoSeconds(rangeStart.toEpochMilli()),
+            convertToNanos == false ? rangeEnd.toEpochMilli() : DateUtils.toNanoSeconds(rangeEnd.toEpochMilli()) };
```

Comment:
Optional: maybe swap the value arguments of the ternary operator for better legibility (i.e. test against `true`).